### ISSUE TITLE
How to handle a pipeline for a hierarchical setup?

### ISSUE THREAD

**Issue by ManuelMartinG** [onnx/sklearn-onnx#627]
Hi,

I'me currently dealing with the following problem. I have a hierarchy of labels in the form:
```python
labels = {
"parent1": 
{"child1", "child2"}, 
"parent2": {"child1", "child2", "child3"}
}
```

My pipeline consists on a first parents classifier model and then I have a classifier for each parent's child labels. I will put it as simple as I can:
```python
class ParentLabelsModel(BaseEstimator, TransformerMixin):
    def __init__(self, base_estimator):
        TransformerMixin.__init__(self)
        BaseEstimator.__init__(self)
        self.base_estimator = base_estimator()

    def fit(self, X, y, *args):
        self.base_estimator.fit(X, y)

    def fit_transform(self, X, y, *args):
        self.base_estimator.fit(X, y)
        return X

    def transform(self, X, *args):
        y = self.base_estimator.predict(X)
        return X, y


class ChildLabelsModel(BaseEstimator, ClassifierMixin):
    def __init__(self, base_estimator, unique_parent_labels):
        BaseEstimator.__init__(self)
        self.unique_parent_labels = unique_parent_labels
        self.child_estimators = {p: base_estimator()
                                 for p in unique_parent_labels}

    def fit(self, X, y, y_, *args):
        X = X.toarray()
        for p in self.unique_parent_labels:
            child_indices = np.argwhere(np.array(y) == p).flatten()
            x_ = X[child_indices, :]
            y__ = y_[child_indices]
            self.child_estimators[p].fit(x_, y__)
        return self

    def predict(self, X, *args):
        parent_y = X[1][0]
        X = X[0].toarray()
        return (np.array([parent_y]),
                np.array([self.child_estimators[parent_y].predict(X)]))


pipeline = Pipeline([
    ("tfidf1", TfidfVectorizer()),
    ("plm", ParentLabelsModel(base_estimator=SGDClassifier)),
    ("clm", ChildLabelsModel(
        base_estimator=SGDClassifier,
        unique_parent_labels=np.unique(ytrain))
    )
])
```

As you can see, the first `ParentsLabelsModel` is a transformer, so that I can pass the predicted parent labels to select the specific `child_estimator` (there is one classifier per parent label).

I've been trying to convert both `ParentLabelsModel` and `ChildLabelsModel` but quite unsuccessfully. Before pasting the code i've written for the `update_registered_converter` functions, I would like to check if you have any opinion or guidelines to approach this conversion.

Thanks!

**Comment by xadupre** [comment]
Hi,

About your first model, you may have a look into this code: https://github.com/sdpython/mlprodict/blob/master/mlprodict/onnx_conv/operator_converters/conv_transfer_transformer.py. It implements a converter for this model: https://github.com/sdpython/mlinsights/blob/master/mlinsights/mlmodel/transfer_transformer.py which does a similar thing to what you do. However, returning X, y in a transform is not handled in sklearn-onnx so I don't know how it would behave. I would probably return only y and use a ColumnTransformer with two transformers, yours and one replicating the input. sklearn-onnx would work on such a pipeline. 

The same remark applies on the second class which does not really follow scikit-learn API as method predict returns two outputs. I would probably merge the two classes into one and write a single converter for it.

sklearn-onnx handles the pipeline logic for you but every piece must follow scikit-learn API, that means predict, predict_proba, decision_function only returns one output. I hope this answer helps you.

Xavier

**Comment by ManuelMartinG** [comment]
Thanks a lot for your response @xadupre !

Regarding returning X and y in the transform, I knew it wasn't very scikit-learn-friendly but I accomodated the inputs for the `ChildLabelsModel` so that it could handle the incoming tuple (X, y) (see where I do `parent_y = X[1][0]`). I will have a deep look at the references you've showed though. I think I would be able to make this model in this way.

Nevertheless, for the second model, the complex part that I haven't been able to figure out is how to mimic the `self.child_estimators[parent_y].predict(X)` into ONNX. Essentially, that line is a sub-model selector depending on the incoming `parent` label. I have tried to define this logic using `OnnxIf` and `OnnxEqual` but I don't know how to specify the 'else_branch` and `then_branch` arguments... and unfortunately not been able to find any example where this operators are used with skl2onnx. Are those operators the right choice to create decision branches? Or should I look into another kind of logic?

Just to give a bit of context, these classes are defined like this just because of an imposed policy from the inference architecture. From your expert opinion, wouldn't it be easier just to transform the first class model and also all the different sub-models into different `InferenceSessions` from each onnx model? We could just wrap the sub-model selector using a simple Python `dict`.

Thanks for your time,

Manuel

**Comment by xadupre** [comment]
You could compute the predict for all of them and do a weighted average with weights in {0, 1}, w_{ij} = 1 means observations is predicted by child_estimator[j]. It is not the most efficient, and really not efficient if the number of child_estimator is high but it works. There are also two ways of doing a loop in ONNX, operators Loop and Scan but they are not always the most efficient way.

**Comment by ManuelMartinG** [comment]
Ok! That would be workaround. But in the end, we are executing inference over all of the child_estimators, which in my use case could be >30 because there are >30 parent labels. I only want one child_estimator to make inference as it would be the one selected by the `parent_label`.

To sum up and close this issue, what you're saying is that **there's no practical way to embed a decision node inside the ONNX graph so that I can execute just the child_estimator that the `parent_label` says?**. If so, how difficult would that be to implement? (thinking now to contribute)

Thanks for your thoughts! Hope that at least this thread serves well to `sklearn-onnx` users :)

**Comment by xadupre** [comment]
It is possible to that with ONNX but it is not always the most efficient way depending on what you want to do (batch size, number of disinct child_estimator). The ONNX graph can be complex.

The converter for the IsolationForest is one of the more complex converter I wrote (https://github.com/onnx/sklearn-onnx/blob/master/skl2onnx/operator_converters/isolation_forest.py, see [IsolationForest ](http://www.xavierdupre.fr/app/mlprodict/helpsphinx/skl_converters/visual-ensemble-022.html?highlight=isolationforest) for a visual representation of the ONNX graph).

I'm currently thinking about a way to write converters using numpy API. It would be a lot easier to handle complex converters.

**Comment by ManuelMartinG** [comment]
That's a scary graph! My onnx graph for these models was getting enormous too... I achieved getting all result labels and probabilities from each `child_estimator` at the end of the graph with an `apply_concat` operation but again, wasn't optimal due to not having the decision branch to select the only one I was interested to make inference. 

This model is thought to be deployed as a web service for 1-observation inference per request so I wasn't considering batch inference. But the execution efficiency is something we consider very important.

I think I will convert each child_estimator in a different ONNX model, independently, which is straightforward and don't need explicit converters. Then, I'll make the routing for the model selector with Python.

That numpy idea sounds exciting! It's been nice digging deep these days through this library. I'll probably start using it quite heavily in the upcoming months.

Thanks again!

**Comment by xadupre** [comment]
I made a kind of proof of concept about the numpy API for ONNX. It works with decorators. I also wrote a [tutorial](http://www.xavierdupre.fr/app/mlprodict/helpsphinx/tutorial/numpy_api_onnx.html) about it. Here is a current transformer. Method *transform* is implemented with ONNX operators but the syntax is similar to numpy.

```
@onnxsklearn_class("onnx_transform", op_version=13)
class DecorrelateTransformerOnnx(TransformerMixin, BaseEstimator):
    def __init__(self, alpha=0.):
        BaseEstimator.__init__(self)
        TransformerMixin.__init__(self)
        self.alpha = alpha

    def fit(self, X, y=None, sample_weights=None):
        self.pca_ = PCA(X.shape[1])  # pylint: disable=W0201
        self.pca_.fit(X)
        return self

    def onnx_transform(self, X):
        if X.dtype is None:
            raise AssertionError("X.dtype cannot be None.")
        mean = self.pca_.mean_.astype(X.dtype)
        cmp = self.pca_.components_.T.astype(X.dtype)
        return (X - mean) @ cmp
```
